crates/ecs/src/archetypes.rs

```diff
@@ -456,15 +533,36 @@ impl World {
     where
         ComponentType: Debug + Send + Sync + 'static,
     {
+        self.add_components_to_entity(entity, [CastableComponent::new(component)])
+    }
+
+    pub(super) fn add_components_to_entity(
+        &mut self,
+        entity: Entity,
+        components: impl IntoIterator<Item = Box<dyn AnyComponent>>,
+    ) -> WorldResult<()> {
         let source_archetype_index = *self
             .entity_to_archetype_index
             .get(&entity)
             .ok_or(WorldError::EntityDoesNotExist(entity))?;
 
-        let add = ArchetypeMutation::Addition(entity);
+        let components: Vec<_> = components.into_iter().collect();
```

Comment:
The collect is needed so I can create multiple iterators from it. The iterator in the function signature is an owning iterator, meaning it's `Item` is an owned `Box<dyn AnyComponent>`. That's necessary because the function needs to take ownership of the new components to store them in the archetype. I considered making it `Clone` but then every component within it would also have to be `Clone`, which seems a bit too strict. 

If we require all components to be `Clone`, then it would probably be possible to avoid this allocation by cloning the iterator. What do you think?

Reply:
Ok fair enough 👍